### DISCUSSION TITLE
Anyone knowing how to initialize an Array Juce container?

### DISCUSSION THREAD

**Discussion by dobo365** [RomanKubiak/ctrlr#397] [Q&A] [answered]
I need to retrieve a list of files from a directory.
This is done by applying the findChildFiles() function on a File object and it returns a Juce Array container.
I don't know the syntax to define this Array object which is different than a classical Lua array.
dD:findChildFiles(2,false,"*.txt") should return the Array of txt files (a list of File objects) in the dD directory

Tried without success:
aa = Array()
aa = Array<File>

I don't want to use OS specific commands (dir, ls...)

Thx in advance for your help!

**Comment by damiensellier** [comment]
https://docs.juce.com/master/classFile.html#a5d2231468698548f2edde113235b3001


It reminds me of a StringAray to me. Like the one returned when you request the MidiInputs list (i struggled with): 

```
saMidiInputDevices = StringArray()
	for i=0,(utils.getMidiInputDevices():size()) do
		saMidiInputDevices:set(i, utils.getMidiInputDevices():get(i))
		
		if utils.getMidiInputDevices():get(i) == myInputDevice then
            deviceIdInput = i
        end -- end if
end
```


I don't know if it's useful but I remind that you have to define an array beforehand, then you get the number of child with getNumChildren()  and populate the array with the childs with getChild(i).

There's a part in the source code in c++ where it uses the getChild(i) at line82 and below:
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Core/CtrlrPanel/CtrlrPanelResourceManager.cpp





This one is deprecated and shows the JUCE to LUA function findChildFiles at line34:
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/Deprecated/CtrlrLuaFile.cpp

This one is up to date and shows findChildFiles at line11:
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/JuceClasses/LFile.h

**Comment by dobo365** [comment]
Thank you Damien.
In fact I tried StringArray() before posting and it works.
So 
a = StringArray() is working
a = Array() is not working

a = {}
a = dD:findChildFiles(2,false,"*.txt") gives
ERROR: No matching overload found, candidates:
void findChildFiles(LFile const&,luabind::object const&,int,bool,String)
dD is a File object.

In LFile.h, it is showing that a table is good
void findChildFiles (luabind::object const& table, int whatToLookFor, bool searchRecursively, const String wildcardPattern) const;

**Pretty sure it is just a syntax thing...**

dD:getNumberOfChildFiles(2,"*.txt") works fine

**Comment by dnaldoog** [comment]
The first argument to the function actually points to a lua table through luabind. The function returns void. The function then fills the lua table with objects of type File (child files):

<pre style='color:red'>
void LFile::findChildFiles (luabind::object const& table, int whatToLookFor, bool searchRecursively, const String wildcardPattern) const
{
	Array<File> files;
	File::findChildFiles (files, whatToLookFor, searchRecursively, wildcardPattern);

	for (int i=1; i<=files.size(); i++)
	{
		table[i] = files[i-1];
	}
}
</pre>

... so in lua you get;

<pre>
t = {}
f = File ("c:\\devel")
f:findChildFiles (t, File.findDirectories, true, "*")
for i,v in ipairs(t) do
	console (v:getFullPathName())
end
</pre>

This is Atom's [code](https://ctrlr.org/forums/search/findChildFiles/). The directory to search is hardcoded. See solution below.

**Comment by dnaldoog** [comment, marked as answer]
This is how would you do it with Directory selection:

<pre>
readFile = function()
    local f = utils.getDirectoryWindow("Search for Directory", File(''))
    local t = {}
    f:findChildFiles(t, 2, false, "*.txt") 
    for i, v in ipairs(t) do
        console(v:getFullPathName())
    end --print each found element in the table
end
</pre>

[findChildFiles_1_0_2021-12-26_18-00.zip](https://github.com/RomanKubiak/ctrlr/files/7776944/findChildFiles_1_0_2021-12-26_18-00.zip)

**Reply by dobo365** [reply]
Thx man!
I knew about utils.getDirectoryWindow() but you answered yourself faster ;-)

To continue on my example above, the problem was indeed a syntax issue and it is needed to write 
dD:findChildFiles(a, 2,false,"*.txt") 
instead of 
a = dD:findChildFiles(2,false,"*.txt")

**Reply by dnaldoog** [reply]
Yes that was a new command for me as was `findChildFiles`! 😃

**Reply by dobo365** [reply]
Worth exploring this one: [https://docs.juce.com/master/classFile.html](https://docs.juce.com/master/classFile.html)
